resources/scripts/ttx_diff.py

```diff
@@ -0,0 +1,157 @@
+#!/usr/bin/env python3
+"""Helper for comparing fontc (Rust) vs fontmake (Python) binaries.
+
+Turns each into ttx, eliminates expected sources of difference and prints
+a brief summary of the result.
+
+fontmake should be installed in an active virtual environment.
+
+Usage:
+    # rebuild with fontmake and fontc and compare
+    python resources/scripts/ttx_diff.py ../OswaldFont/sources/Oswald.glyphs
+    # rebuild the fontc copy, reuse the prior fontmake copy (if present), and compare
+    python resources/scripts/ttx_diff.py --rebuild fontc ../OswaldFont/sources/Oswald.glyphs
+"""
+
+from absl import app
+from absl import flags
+from lxml import etree
+from pathlib import Path
+import shutil
+import subprocess
+import sys
+
+
+FLAGS = flags.FLAGS
+
+
+flags.DEFINE_enum('rebuild', 'both', ['both', 'fontc', 'fontmake'], 'Which compilers to rebuild with if the output appears to already exist')
+
+
+def run(cmd, **kwargs):
+    cmd_string = " ".join(cmd)
+    print(f"Running {cmd_string}")
+    subprocess.run(cmd, text=True, check=True, **kwargs)
+
+
+def ttx(font_file):
+    ttx_file = font_file.with_suffix(".ttx")
+    cmd = [
+        "ttx",
+        "-o",
+        str(ttx_file),
+        str(font_file),
+    ]
+    run(cmd)
+    return ttx_file
+
+
+def build(cmd, build_tool, ttf_find_fn, **kwargs):
+    try_skip = FLAGS.rebuild not in [build_tool, "both"]
+    ttfs = ttf_find_fn()
+    if try_skip and len(ttfs) == 1:
+        ttx_file = ttfs[0].with_suffix(".ttx")
+        if ttx_file.is_file():
+            print(f"skipping {build_tool}")
+            return ttx_file 
+    run(cmd, **kwargs)
+    ttfs = ttf_find_fn()
+    assert len(ttfs) == 1, ttfs
+    return ttx(ttfs[0])
+
+
+def build_fontc(source, build_dir):    
+    cmd = [
+        "cargo",
+        "run",
+        "-p",
+        "fontc",
+        "--",
+        "--source",
+        str(source),
+        "--build-dir",
+        str(build_dir),
+    ]
+    return build(cmd, "fontc", lambda: (build_dir / "font.ttf",))
+
+
+def build_fontmake(source, build_dir):
+    cmd = [
+        "fontmake",
+        "-o",
+        "variable",
+        "--no-production-names",
+        "--keep-direction",
+        "--filter",
+        "FlattenComponentsFilter",
+        "--filter",
+        "DecomposeTransformedComponentsFilter",
```

Comment:
note that this optional filter is more eager than our current implementation in fontc. This one will decompose any transformed components anywhere, not just those components that have different 2x2 in different masters like we currently do. We could add an option to do the former (off by default because fontmake doesn't enable this by default).

Reply:
Good point, moving to a standalone issue.

Reply:
https://github.com/googlefonts/fontmake-rs/issues/291